File: bgpx/session.py
"""BGP session FSM — always listens AND actively connects; first to succeed wins."""

import asyncio
import logging
from dataclasses import dataclass

from bgpx.constants import (
    BGP_HEADER_LEN,
    MSG_OPEN, MSG_UPDATE, MSG_NOTIFICATION, MSG_KEEPALIVE,
)
from bgpx.events import EventBus
from bgpx.message.builder import build_open, build_keepalive
from bgpx.message.parser import parse_header, parse_open, parse_update_details
from bgpx.rib import FlowspecRIB

log = logging.getLogger(__name__)

IDLE           = "IDLE"
CONNECT        = "CONNECT"
OPEN_SENT      = "OPEN_SENT"
OPEN_CONFIRMED = "OPEN_CONFIRMED"
ESTABLISHED    = "ESTABLISHED"
LISTENING      = "LISTENING"


@dataclass
class SessionConfig:
    local_as:        int
    router_id:       str
    peer_ip:         str
    peer_as:         int
    hold_time:       int           = 90
    reconnect_delay: int           = 5
    connect_timeout: float         = 5.0
    active_retry_delay: float      = 1.0
    listen_port:     int           = 179
    json_output:     str | None = None


@dataclass
class _Connection:
    reader:  asyncio.StreamReader
    writer:  asyncio.StreamWriter
    role:    str
    peer_ip: str


class BGPSession:
    def __init__(
        self,
        config: SessionConfig,
        rib:    FlowspecRIB,
        events: EventBus | None = None,
    ):
        self.config     = config
        self.rib        = rib
        self._events    = events
        self._state     = IDLE
        self._peer_info: dict = {}
        self._hold_time = config.hold_time
        self._ka_task:  asyncio.Task | None = None
        self._running   = True

        # Queue fed by the persistent passive server
        # Size=1 so we don't accumulate stale connections
        self._incoming: asyncio.Queue[_Connection] = asyncio.Queue(maxsize=1)
        self._incoming_event = asyncio.Event()

    # ── Public ────────────────────────────────────────────────────────────────

    @property
    def state(self) -> str:
        return self._state

    @property
    def peer_info(self) -> dict:
        return self._peer_info

    async def run(self):
        """
        Two permanent background tasks run concurrently:
          1. _passive_server  — listens on listen_port forever, feeds _incoming queue
          2. _active_loop     — repeatedly races active-connect vs incoming connection
        """
        server_task = asyncio.create_task(self._passive_server())
        try:
            await self._session_loop()
        finally:
            server_task.cancel()
            try:
                await server_task
            except (asyncio.CancelledError, Exception):
                pass

    def stop(self):
        self._running = False

    # ── Persistent passive server ─────────────────────────────────────────────

    async def _passive_server(self):
        """Always listen on listen_port. Put peer connections in the queue."""
        port = self.config.listen_port
        try:
            server = await asyncio.start_server(
                self._on_incoming, "0.0.0.0", port,
            )
        except PermissionError:
            self._emit(
                "session",
                f"Cannot bind port {port} — passive disabled "
                f"(run with sudo or: sudo setcap cap_net_bind_service+ep $(readlink -f $(which python3)))",
                level="warning",
            )
            log.warning(f"Cannot bind port {port} — passive mode disabled")
            return
        except asyncio.CancelledError:
            return

        self._emit("session", f"Passive listener started on 0.0.0.0:{port}")
        log.info(f"Passive listener on 0.0.0.0:{port}")

        try:
            async with server:
                await server.serve_forever()
        except asyncio.CancelledError:
            pass

    async def _on_incoming(
        self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter
    ):
        peer_ip = writer.get_extra_info("peername", ("unknown",))[0]

        if peer_ip != self.config.peer_ip:
            log.warning(f"Passive: rejected connection from {peer_ip} (not peer)")
            await self._close_writer(writer)
            return

        log.info(f"Passive: incoming connection from {peer_ip}")
        self._emit("session", f"Passive: incoming connection from {peer_ip}")

        if self._state not in (IDLE, CONNECT):
            log.info(
                f"Passive: dropping duplicate connection from {peer_ip} "
                f"while state={self._state}"
            )
            self._emit(
                "session",
                f"Passive: duplicate connection from {peer_ip} dropped "
                f"while {self._state}",
                level="warning",
            )
            await self._close_writer(writer)
            return

        try:
            # Drop if queue is full (session already in progress)
            self._incoming.put_nowait(
                _Connection(reader, writer, "passive", peer_ip)
            )
            self._incoming_event.set()
        except asyncio.QueueFull:
            log.debug("Passive: session already in progress, dropping incoming connection")
            await self._close_writer(writer)

    # ── Session loop ──────────────────────────────────────────────────────────

    async def _session_loop(self):
        """Outer loop: race active vs passive, run session, repeat."""
        while self._running:
            reader, writer = await self._race_connect()
            if reader is None:
                break

            try:
                await self._run_session(reader, writer)
            except asyncio.CancelledError:
                raise
            except Exception as e:
                log.error(f"Session error: {e}")
                self._emit("error", f"Session error: {e}", level="error", error=str(e))
            finally:
                await self._teardown()
                try:
                    writer.close()
                    await writer.wait_closed()
                except Exception:
                    pass

            if self._running:
                delay = self.config.reconnect_delay
                await self._wait_reconnect_delay(delay)

    async def _race_connect(self) -> tuple:
        """Race active-connect vs passive-accept. Return (reader, writer) of winner."""
        self._set_state(CONNECT)

        try:
            conn = self._incoming.get_nowait()
        except asyncio.QueueEmpty:
            conn = None

        if conn:
            if self._incoming.empty():
                self._incoming_event.clear()
            self._emit(
                "session",
                f"Passive: using queued connection from {conn.peer_ip}",
            )
            return conn.reader, conn.writer

        self._emit(
            "session",
            f"Racing: active→{self.config.peer_ip}:179 vs passive←:{self.config.listen_port}",
        )

        active_task   = asyncio.create_task(self._active_attempts())
        incoming_task = asyncio.create_task(self._incoming.get())

        try:
            done, pending = await asyncio.wait(
                [active_task, incoming_task],
                return_when=asyncio.FIRST_COMPLETED,
            )
        except asyncio.CancelledError:
            active_task.cancel()
            incoming_task.cancel()
            for task in (active_task, incoming_task):
                try:
                    await task
                except (asyncio.CancelledError, Exception):
                    pass
            return None, None

        await asyncio.sleep(0)
        ready_pending = {t for t in pending if t.done()}
        done.update(ready_pending)
        pending.difference_update(ready_pending)

        for task in pending:
            task.cancel()
            try:
                await task
            except (asyncio.CancelledError, Exception):
                pass

        candidates: list[_Connection] = []
        for task in done:
            if task.cancelled():
                continue
            if task.exception():
                continue
            candidates.append(task.result())

        if not candidates:
            return None, None

        selected = self._select_connection(candidates)
        for conn in candidates:
            if conn is not selected:
                await self._close_connection(conn)

        return selected.reader, selected.writer

    async def _active_attempts(self) -> tuple:
        """Keep trying to connect to peer:179 until success."""
        attempt = 0
        while True:
            attempt += 1
            try:
                log.debug(f"Active attempt #{attempt} → {self.config.peer_ip}:179")
                reader, writer = await asyncio.wait_for(
                    asyncio.open_connection(self.config.peer_ip, 179),
                    timeout=self.config.connect_timeout,
                )
                peer = writer.get_extra_info("peername")
                log.info(f"Active: TCP connected to {peer}")
                self._emit("session", f"Active: TCP connected to {peer}")
                peer_ip = peer[0] if peer else self.config.peer_ip
                return _Connection(reader, writer, "active", peer_ip)
            except asyncio.CancelledError:
                raise
            except asyncio.TimeoutError:
                log.debug(
                    f"Active connect timed out after {self.config.connect_timeout}s "
                    f"— retry in {self.config.active_retry_delay}s"
                )
                await asyncio.sleep(self.config.active_retry_delay)
            except Exception as e:
                log.debug(
                    f"Active connect failed: {e} "
                    f"— retry in {self.config.active_retry_delay}s"
                )
                await asyncio.sleep(self.config.active_retry_delay)

    # ── BGP session ───────────────────────────────────────────────────────────

    async def _run_session(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
        cfg = self.config
        self._set_state(OPEN_SENT)
        await self._send(writer, build_open(cfg.local_as, cfg.hold_time, cfg.router_id))
        log.info("Sent OPEN")
        self._emit("session", "Sent OPEN")
        await self._rx_loop(reader, writer)

    async def _rx_loop(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
        while True:
            timeout = self._hold_time if self._hold_time > 0 else None
            try:
                hdr = await asyncio.wait_for(
                    reader.readexactly(BGP_HEADER_LEN), timeout=timeout
                )
            except asyncio.TimeoutError:
                raise ConnectionError(f"Hold timer expired ({self._hold_time}s) — no message from peer")
            msg_type, body_len = parse_header(hdr)
            body = await reader.readexactly(body_len) if body_len > 0 else b''
            await self._dispatch(msg_type, body, writer)

    async def _dispatch(self, msg_type: int, body: bytes, writer: asyncio.StreamWriter):
        if msg_type == MSG_OPEN:
            info = parse_open(body)
            if info["peer_as"] != self.config.peer_as:
                raise ConnectionError(
                    f"Peer ASN mismatch: expected {self.config.peer_as}, got {info['peer_as']}"
                )
            self._peer_info = info
            self._hold_time = min(self.config.hold_time, info["hold_time"])
            log.info(
                f"Received OPEN  peer_as={info['peer_as']}"
                f"  router_id={info['router_id']}"
                f"  hold_time={self._hold_time}s"
            )
            self._emit(
                "session",
                f"Received OPEN from AS {info['peer_as']} ({info['router_id']})"
                f"  hold_time={self._hold_time}s",
                peer_info=info,
            )
            await self._send(writer, build_keepalive())
            self._set_state(OPEN_CONFIRMED)
            if self._hold_time > 0:
                self._ka_task = asyncio.create_task(self._ka_loop(writer))

        elif msg_type == MSG_KEEPALIVE:
            if self._state == OPEN_CONFIRMED:
                self._set_state(ESTABLISHED)
                log.info("Session ESTABLISHED ✓")
            else:
                log.debug("Received KEEPALIVE")

        elif msg_type == MSG_UPDATE:
            if self._state != ESTABLISHED:
                return
            update = parse_update_details(
                body,
                asn_len=4 if self._peer_info.get("supports_4byte_asn") else 2,
            )
            announce = update["announce"]
            withdraw = update["withdraw"]
            actions = update["actions"]
            path_attributes = update["path_attributes"]
            for afi, routes in announce.items():
                for route in routes:
                    if afi.endswith("-unicast"):
                        as_path, communities = _unicast_attributes(path_attributes)
                        route_id = self.rib.add_unicast(
                            afi=afi,
                            prefix=route["prefix"],
                            peer=self.config.peer_ip,
                            next_hop=route.get("next_hop", ""),
                            as_path=as_path,
                            communities=communities,
                            path_attributes=path_attributes,
                        )
                        extra = {
                            "family": "unicast",
                            "prefix": route["prefix"],
                            "next_hop": route.get("next_hop", ""),
                            "as_path": as_path,
                            "communities": communities,
                        }
                    else:
                        route_id = self.rib.add_flowspec(
                            afi, route, actions, self.config.peer_ip,
                            path_attributes=path_attributes,
                        )
                        extra = {
                            "family": "flowspec",
                            "match": route,
                            "actions": actions,
                        }
                    self._emit(
                        "announce", f"ANNOUNCE {afi}",
                        level="update",
                        afi=afi, route_id=route_id,
                        path_attributes=path_attributes,
                        peer=self.config.peer_ip,
                        **extra,
                    )
            for afi, routes in withdraw.items():
                for route in routes:
                    if afi.endswith("-unicast"):
                        route_id = self.rib.remove_unicast(
                            afi, route["prefix"], self.config.peer_ip
                        )
                        extra = {
                            "family": "unicast",
                            "prefix": route["prefix"],
                        }
                    else:
                        route_id = self.rib.remove_flowspec(
                            afi, route, self.config.peer_ip
                        )
                        extra = {
                            "family": "flowspec",
                            "match": route,
                        }
                    self._emit(
                        "withdraw", f"WITHDRAW {afi}",
                        level="update",
                        afi=afi, route_id=route_id,
                        path_attributes=path_attributes,
                        peer=self.config.peer_ip,
                        **extra,
                    )

        elif msg_type == MSG_NOTIFICATION:
            error_code = body[0] if body else 0
            subcode    = body[1] if len(body) > 1 else 0
            log.error(f"NOTIFICATION error={error_code} subcode={subcode}")
            self._emit(
                "error", f"NOTIFICATION error={error_code} subcode={subcode}",
                level="error", error_code=error_code, subcode=subcode,
            )
            raise ConnectionError(f"BGP NOTIFICATION {error_code}/{subcode}")

        else:
            log.warning(f"Unknown message type {msg_type}")

    # ── Helpers ───────────────────────────────────────────────────────────────

    def _emit(self, event_type: str, message: str, level: str = "info", **kw):
        if self._events:
            self._events.emit(event_type, level, message, **kw)

    def _set_state(self, state: str):
        self._state = state
        self._emit("session", f"State → {state}", state=state)

    async def _teardown(self):
        prev            = self._state
        self._state     = IDLE
        self._peer_info = {}
        if self._ka_task:
            self._ka_task.cancel()
            try:
                await self._ka_task
            except (asyncio.CancelledError, Exception):
                pass
            self._ka_task = None
        self.rib.clear_peer(self.config.peer_ip)
        await self._discard_queued_incoming()
        if prev not in (IDLE, LISTENING, CONNECT):
            self._emit("session", "Disconnected", level="warning", state=IDLE)

    async def _send(self, writer: asyncio.StreamWriter, data: bytes):
        writer.write(data)
        await writer.drain()

    async def _ka_loop(self, writer: asyncio.StreamWriter):
        interval = max(1, self._hold_time // 3)
        while True:
            await asyncio.sleep(interval)
            await self._send(writer, build_keepalive())
            log.debug("Sent KEEPALIVE")

    def _select_connection(self, candidates: list[_Connection]) -> _Connection:
        for conn in candidates:
            if conn.role == "passive":
                return conn
        return candidates[0]

    async def _discard_queued_incoming(self):
        while True:
            try:
                conn = self._incoming.get_nowait()
            except asyncio.QueueEmpty:
                self._incoming_event.clear()
                return
            await self._close_connection(conn)

    async def _wait_reconnect_delay(self, delay: int | float):
        if delay <= 0:
            return
        if not self._incoming.empty():
            self._emit(
                "session",
                "Passive connection already queued; reconnecting immediately",
            )
            return

        self._emit(
            "session",
            f"Reconnecting in {delay:g}s …",
            level="warning",
        )
        try:
            await asyncio.wait_for(self._incoming_event.wait(), timeout=delay)
        except asyncio.TimeoutError:
            return

        if self._incoming.empty():
            self._incoming_event.clear()
            return
        self._emit(
            "session",
            "Passive connection arrived; reconnecting immediately",
        )

    async def _close_connection(self, conn: _Connection):
        await self._close_writer(conn.writer)

    async def _close_writer(self, writer: asyncio.StreamWriter):
        writer.close()
        try:
            await writer.wait_closed()
        except Exception:
            pass


def _unicast_attributes(path_attributes: list[dict]) -> tuple[list[int], list[str]]:
    as_path: list[int] = []
    as4_path: list[int] = []
    communities: list[str] = []

    for attribute in path_attributes:
        value = attribute.get("value")
        if attribute["name"] in ("AS_PATH", "AS4_PATH") and isinstance(value, list):
            flattened = [
                asn
                for segment in value
                for asn in segment.get("asns", [])
            ]
            if attribute["name"] == "AS4_PATH":
                as4_path = flattened
            else:
                as_path = flattened
        elif attribute["name"] in ("COMMUNITIES", "LARGE_COMMUNITIES"):
            if isinstance(value, list):
                communities.extend(value)

    if as4_path:
        as_path = as_path[:-len(as4_path)] + as4_path
    return as_path, communities

File: src/session.rs
use crate::{app::Shared, wire};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::{net::Ipv4Addr, time::Duration};
use tokio::{
    io::{AsyncReadExt, AsyncWriteExt},
    net::{TcpListener, TcpStream},
    sync::oneshot,
    time::{sleep, timeout, Instant},
};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Config {
    pub local_as: u32,
    pub router_id: Ipv4Addr,
    pub peer_ip: Ipv4Addr,
    pub peer_as: u32,
    pub hold_time: u16,
    pub reconnect_delay: f64,
    pub connect_timeout: f64,
    pub active_retry_delay: f64,
    pub listen_port: u16,
    pub json_output: Option<String>,
}
impl Config {
    pub fn from_value(mut value: Value) -> Result<Self, String> {
        let map = value.as_object_mut().ok_or("Config must be an object")?;
        for (key, default) in [
            ("hold_time", json!(90)),
            ("reconnect_delay", json!(5)),
            ("connect_timeout", json!(5.0)),
            ("active_retry_delay", json!(1.0)),
            ("listen_port", json!(179)),
        ] {
            map.entry(key).or_insert(default);
        }
        for key in [
            "local_as",
            "peer_as",
            "hold_time",
            "reconnect_delay",
            "connect_timeout",
            "active_retry_delay",
            "listen_port",
        ] {
            if let Some(s) = map.get(key).and_then(Value::as_str) {
                let parsed: Value =
                    serde_json::from_str(s).map_err(|_| format!("Invalid {key}"))?;
                if !parsed.is_number() {
                    return Err(format!("Invalid {key}"));
                }
                map.insert(key.into(), parsed);
            }
        }
        if map.get("json_output") == Some(&json!("")) {
            map.insert("json_output".into(), Value::Null);
        }
        let config: Self = serde_json::from_value(value).map_err(|e| e.to_string())?;
        config.validate()?;
        Ok(config)
    }
    pub fn validate(&self) -> Result<(), String> {
        if self.local_as == 0 || self.peer_as == 0 {
            return Err("AS numbers must be positive".into());
        }
        if self.hold_time != 0 && self.hold_time < 3 {
            return Err("Hold time must be zero or at least 3 seconds".into());
        }
        if self.listen_port == 0 {
            return Err("Listen port must be positive".into());
        }
        for (name, value, allow_zero) in [
            ("reconnect_delay", self.reconnect_delay, true),
            ("connect_timeout", self.connect_timeout, false),
            ("active_retry_delay", self.active_retry_delay, false),
        ] {
            if !value.is_finite() || value < 0.0 || (!allow_zero && value == 0.0) || value > 86400.0
            {
                return Err(format!(
                    "Invalid {name}: expected {}..86400 seconds",
                    if allow_zero { "0" } else { "positive" }
                ));
            }
        }
        Ok(())
    }
}

async fn incoming(listener: &Option<TcpListener>, config: &Config) -> TcpStream {
    let Some(listener) = listener else {
        return std::future::pending().await;
    };
    loop {
        match listener.accept().await {
            Ok((stream, peer)) if peer.ip() == config.peer_ip => return stream,
            Ok(_) => {}
            Err(_) => sleep(Duration::from_millis(100)).await,
        }
    }
}
async fn active(config: &Config) -> TcpStream {
    loop {
        if let Ok(Ok(stream)) = timeout(
            Duration::from_secs_f64(config.connect_timeout),
            TcpStream::connect((config.peer_ip, 179)),
        )
        .await
        {
            return stream;
        }
        sleep(Duration::from_secs_f64(config.active_retry_delay)).await;
    }
}

pub async fn run(app: Shared, config: Config, mut stop: oneshot::Receiver<()>) {
    let listener = match TcpListener::bind((Ipv4Addr::UNSPECIFIED, config.listen_port)).await {
        Ok(l) => {
            app.emit(
                "session",
                "info",
                format!("Passive listener on :{}", config.listen_port),
                json!({}),
            );
            Some(l)
        }
        Err(e) => {
            app.emit(
                "session",
                "warning",
                format!("Passive listener disabled: {e}"),
                json!({}),
            );
            None
        }
    };
    let mut queued = None;
    loop {
        app.set_state("CONNECT");
        let stream = if let Some(stream) = queued.take() {
            stream
        } else {
            tokio::select! {biased;
                _=&mut stop=>break,
                stream=incoming(&listener,&config)=>stream,
                stream=active(&config)=>stream,
            }
        };
        let result = {
            let session = connected(&app, &config, stream);
            tokio::pin!(session);
            loop {
                tokio::select! {biased;
                    _=&mut stop=>break None,
                    result=&mut session=>break Some(result),
                    duplicate=incoming(&listener,&config)=>{
                        drop(duplicate);
                        app.emit("session","warning","Duplicate passive connection dropped",json!({}));
                    }
                }
            }
        };
        {
            let mut data = app.data.lock().unwrap();
            data.clear();
            data.peer_info = json!({});
        }
        app.set_state("IDLE");
        let Some(result) = result else { break };
        if let Err(e) = result {
            app.emit("error", "error", e, json!({}));
        }
        tokio::select! {biased;
            _=&mut stop=>break,
            stream=incoming(&listener,&config)=>queued=Some(stream),
            _=sleep(Duration::from_secs_f64(config.reconnect_delay))=>{},
        }
    }
    {
        let mut data = app.data.lock().unwrap();
        data.running = false;
        data.state = "IDLE";
        data.clear();
        data.peer_info = json!({});
    }
    if let Err(e) = app.flush(true) {
        app.emit("error", "error", e, json!({}));
    }
}

async fn read_message(
    reader: &mut tokio::net::tcp::OwnedReadHalf,
) -> Result<(u8, Vec<u8>), String> {
    let mut header = [0; 19];
    reader
        .read_exact(&mut header)
        .await
        .map_err(|e| e.to_string())?;
    let (kind, len) = wire::header(&header)?;
    let mut body = vec![0; len];
    reader
        .read_exact(&mut body)
        .await
        .map_err(|e| e.to_string())?;
    Ok((kind, body))
}

async fn connected(app: &Shared, config: &Config, stream: TcpStream) -> Result<(), String> {
    stream.set_nodelay(true).map_err(|e| e.to_string())?;
    let (mut reader, mut writer) = stream.into_split();
    app.set_state("OPEN_SENT");
    writer
        .write_all(&wire::build_open(
            config.local_as,
            config.hold_time,
            config.router_id,
        ))
        .await
        .map_err(|e| e.to_string())?;
    let mut hold = config.hold_time;
    let mut received_open = false;
    let mut established = false;
    let mut asn_len = 2;
    let mut deadline =
        Instant::now() + Duration::from_secs(if hold == 0 { 86400 * 365 } else { hold as u64 });
    let mut next_keepalive = Instant::now() + Duration::from_secs(86400 * 365);
    loop {
        // Keep the frame read alive while sending keepalives; read_exact is not cancellation safe.
        let frame = read_message(&mut reader);
        tokio::pin!(frame);
        let (kind, body) = loop {
            tokio::select! {
                result=&mut frame=>break result?,
                _=tokio::time::sleep_until(deadline),if hold>0=>return Err("Hold timer expired".into()),
                _=tokio::time::sleep_until(next_keepalive),if received_open&&hold>0=>{
                    timeout(Duration::from_secs(hold as u64),writer.write_all(&wire::message(4,&[]))).await.map_err(|_|"Keepalive write timed out")?.map_err(|e|e.to_string())?;
                    next_keepalive=Instant::now()+Duration::from_secs((hold/3).max(1) as u64);
                }
            }
        };
        match kind {
            1 if !received_open => {
                let info = wire::open(&body)?;
                if info["peer_as"] != config.peer_as {
                    return Err("Peer ASN mismatch".into());
                }
                hold = hold.min(info["hold_time"].as_u64().unwrap() as u16);
                if info["supports_4byte_asn"] == true {
                    asn_len = 4;
                }
                app.data.lock().unwrap().peer_info = info.clone();
                app.emit(
                    "session",
                    "info",
                    "Received OPEN",
                    json!({"peer_info":info}),
                );
                writer
                    .write_all(&wire::message(4, &[]))
                    .await
                    .map_err(|e| e.to_string())?;
                received_open = true;
                app.set_state("OPEN_CONFIRMED");
                next_keepalive = Instant::now() + Duration::from_secs((hold / 3).max(1) as u64);
            }
            4 if received_open && body.is_empty() => {
                if !established {
                    established = true;
                    app.set_state("ESTABLISHED");
                }
            }
            2 if established => {
                app.apply_update(wire::update(&body, asn_len)?, &config.peer_ip.to_string())
            }
            3 => {
                return Err(format!(
                    "BGP NOTIFICATION {}/{}",
                    body.first().unwrap_or(&0),
                    body.get(1).unwrap_or(&0)
                ))
            }
            _ => return Err(format!("Unexpected BGP message {kind}")),
        }
        if hold > 0 {
            deadline = Instant::now() + Duration::from_secs(hold as u64);
        }
    }
}
